Compile resx file into binary resource file for use in AssemblyAction

jrbeverly opened the pull request:
Compiles a single resx file into a common language runtime binary resource file.

Example usage:
```bazel
csharp_resx(
    name = "hello.resources.mystrings",
    src = "Strings.resx",
    identifier = "Hello.Strings.resources",
)
```

There a couple of things I'll mention here:

- ResX files have a lot of cases that I think should be tested (multiple resx files, embedded files, identifier/out parameters, etc)
- A target_framework is necessary for the csproj to compile

Resolves #10 , #77

jrbeverly commented on csharp/private/rules/resgen.bzl:
```diff
@@ -0,0 +1,99 @@
+        arguments = [args],
+        mnemonic = "CompileResX",
+        progress_message = "Compiling resx file to binary",
+        env = {
```
Without these 4 environment variables, `dotnet` cli just won't run. @j3parker, do you have any ideas on how I might be able to: 1) not have to specify these 2) specify these for each OS

While they are set to `/root/`, they will fail on the linux & macOS builds (as they wouldn't have permission to create `/root/.dotnet`). Additionally without `DOTNET_CLI_TELEMETRY_OPTOUT=1`, the telemetry message will output a bunch.

jrbeverly replied:
Getting back into this after inspiration, I think the best option for now would be a wrapper over `dotnet` that ensures the appropriate variables are passed in for each execution. This way, we don't have to worry about the `.dotnet` directory of one rule conflicting with that of another.

I looked into other options, but they were mostly dead ends.

omsmith commented on examples/resgen/Strings.Designer.cs:
```diff
@@ -0,0 +1,30 @@
+﻿namespace Hello {
```
Should this have been committed?

jrbeverly replied:
The `.Designer.cs` file? Yes. Without it you would just get a compilation error about `Strings` not existing in the context.

You can generate this file in the compilation process using a `Compile` task on the designer file (e.g. `<Compile Update="Strings.Designer.cs">`).

omsmith commented on examples/resgen/BUILD:
```diff
@@ -0,0 +1,18 @@
+csharp_resx(
+    name = "Hello.Strings",
+    src = "Strings.resx",
+)
```
nit: Missing trailing newlines on a number of files

omsmith commented on csharp/private/rules/resgen.bzl:
```diff
@@ -0,0 +1,99 @@
+# be in the rule bin directory. If we want to get the relative path to the
+# ResX files, we need to get to a junction/symlink for the source. 
+#
+# The junction/symlink that we are using is 4 directories up (\bazel-out\<sandbox>\bin\<name>\)
```
Relying on this seems, likely to break.

jrbeverly replied:
Agree, but I have had difficulty finding a way of configuring relative paths for the `.resx` files that isn't a bit 'hacky'.

Ideally I'd just go with the copy route, but that doesn't handle embedded files like so:

```xml
  <data name="SetupFiles" type="System.Resources.ResXFileRef, System.Windows.Forms">
    <value>Resources\SetupFiles.zip;System.Byte[], mscorlib, Version=4.0.0.0, ...</value>
  </data>
```

jrbeverly commented:
I have split out the work for running `dotnet` in another pull request #102. Using some of the work I did there with runfiles, I will be revising how the resx files are compiled in this pull request. At the moment I use a pathing hack (`../../../../`) to try and find the `resx` file. This only works for files that are one directory below the WORKSPACE (overfitting of the examples). 

The new approach I'll be using the runfiles data to create the csproj file. This will use an absolute path for the resx file (ensuring all embedded files are working as intended), and convert the rule to a macro.

I'll look to raise a pull request tomorrow that uses this new approach for compilation of resx files.

jrbeverly commented:
I have managed to get a solution working with the runfiles (and learned a fair bit about runfiles). You can see a preview of what is looks like [here](https://github.com/Brightspace/rules_csharp/compare/master...jrbeverly:proto/resx-macro-tangent-2?expand=1). I'll be including the learnings from that branch into this one, as that solution is a bit wonky.

I'll be trimming it down and including the changes into this pull request.

jrbeverly commented:
I have simplified the work done down to two rules, and moved the C++ compilation wrapper into the bash wrapper. This wrapper doesn't have as much going on (besides the boilerplate runfiles code).

I did discover something while working on windows, and that is that there exists a segmentation fault in the `dotnetw` wrapper. So I'll creating an issue to try to resolve that. I suspect it is related to the `dotnet_argv` that are passed in from `argv`. I suspect it needs to be copied rather than just reference.

j3parker commented:
Sorry in the delay getting back to this, I'll go through it on the weekend when I have some quiet time :)

jrbeverly commented:
While doing the housekeeping on some of these PRs, I noticed that with the latest release of the dotnet SDK `3.1.100`, it introduced the expectation for `project.assets.json`.

That was failing the resx compilation due to its usage of `dotnet build`. Based on a quick search, I suspect this'll impact the current implementation of resx.

omsmith commented:
:(
